routing/DELEGATED_IPNS_HTTP.md

```diff
@@ -0,0 +1,176 @@
+# Delegated IPNS HTTP API
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+**Author(s)**:
+
+- [Masih H. Derkani](https://github.com/masih)
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+Delegated IPNS is a mechanism for IPFS implementations to offload naming system to another process
+or server. This includes naming resolution as well as publication of new naming records. This
+document describes an HTTP API through which such functionality is facilitated.
+
+## API Specification
+
+The Delegated IPNS HTTP API uses the `application/json` content type by default.
+
+As such, human-readable encodings of types are preferred. This spec may be updated in the future
+with a compact `application/cbor` encoding, in which case compact encodings of the various types
+would be used.
+
+## Common Data Types
+
+### IPNS Record
+
+The following snippet outlines the JSON schema of IPNS records:
+
+```json
+{
+  "Signature": "<signature>",
+  "Payload": {
+    "Value": "<value>",
+    "Sequence": 0,
+    "Validity": {
+      "EOL": {
+        "Timestamp": 0,
+        "AdvisoryTTL": 0
+      }
+    },
+    "PublicKey": "<optional-public-key>",
+    "ExtendedData": {}
+  }
+}
+```
+
+Where:
+
+- `Signature` is the multibase-encoded signature of the sha256 hash of the `Payload` field, signed
+  using the private key that corresponds to the `PublicKey` in the `Payload` if present. And
+  Otherwise, the private key associcated to the IPNS record key. Signing details for specific key
+  types should
+  follow [libp2p/peerid specs](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types),
+  unless stated otherwise.
+- `Payload` is the content of the IPNS record as specified
+  by [IPNS Record](../ipns/IPNS.md#ipns-record) specification:
+- `Value` is the string representation of the IPNS path,
+  e.g. `ipns/{ipns-key}`, `/ipns/example.com`, `/ipfs/baf...`, etc.
+- `Sequence` represents the current version of the record starting from `0`.
+- `Validity` captures the mechanism by which the record is validated. Each validity type reserves a
+  field key under this object.
+  - `EOL` donates that the validity type is EOL, containing:
+    - `Timestamp` represents the time in the future at which the record expires with nanoseconds
+       precision represented as an ASCII string that follows notation
+       from [RFC3339](https://www.ietf.org/rfc/rfc3339.txt).
+    - `AdvisoryTTL` represents an optional field that hints at how long the record should be
+       cached.
+- `PublicKey` represents the optional public key used to sign the record. This field is only
+  required if it cannot be extracted from the IPNS name, e.g. in the case of legacy RSA keys.
```

Comment:
When a future implementer reads this, it may not be obvious why this is important, so let's clarify:

```suggestion
  required if it cannot be extracted from the IPNS name, e.g. in the case of legacy RSA keys.
  The idea here is that IPNS Record should always have all information neccessay
  for signature verification in offline mode, without additional lookups to retrieve the public key.
```